### DISCUSSION TITLE
0.12.0-stable has been released!

### DISCUSSION THREAD

**Discussion by mihe** [godot-jolt/godot-jolt#745] [Announcements]
This release only supports **Godot 4.2** (including 4.2.x).

## Changelog

These are the notable changes that have been made since [0.11.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.11.0-stable) was released. You can also find a list of all the commits [here](https://github.com/godot-jolt/godot-jolt/compare/v0.11.0-stable...v0.12.0-stable).

Any breaking changes are denoted with ⚠️.

### Changed

- ⚠️ Changed so that single-body joints now implicitly sets `node_a` to be the "world node" rather than `node_b`. This diverges from how Godot Physics behaves, but matches how Bullet behaves in Godot 3, and yields more intuitive outcomes for the 6DOF joints.
- ⚠️ Changed `Generic6DOFJoint3D` and `ConeTwistJointImpl3D`, as well as their substitute joints, to use pyramid-shaped angular limits instead of cone-shaped limits, to better match Godot Physics.
- ⚠️ Reversed the direction of the `equilibrium_point` properties for `Generic6DOFJoint3D` and `JoltGeneric6DOFJoint3D`, to match the direction of the angular limits.
- ⚠️ Changed the rotation order of the `equilibrium_point` properties for `Generic6DOFJoint3D` and `JoltGeneric6DOFJoint3D`, from ZXY to XYZ, to match the rotation order of the angular limits.
- Mirrored the way in which linear limits are visualized for `JoltSliderJoint3D` and `JoltGeneric6DOFJoint3D`.

### Added

- Added new project setting, "World Node", for controlling which of the two nodes in a single-body joint becomes the "world node" when omitting one of the nodes. This allows for reverting back to the behavior of Godot Physics if needed, effectively undoing the breaking change mentioned above.
- Added new project setting, "Report All Kinematic Contacts", for allowing `RigidBody3D` frozen with `FREEZE_MODE_KINEMATIC` to report contacts/collisions with other kinematic/static bodies, at a potentially heavy performance/memory cost.
- Added support for using NaN to indicate holes in `HeightMapShape3D`.
- Added support for holes in a non-square `HeightMapShape3D`.

### Fixed

- ⚠️ Fixed issue with non-square `HeightMapShape3D` not using back-face collision.
- Fixed issue where contact shape indices would sometimes always be the same index across all contacts with a particular body.
- Fixed runtime crash when setting the `max_contacts_reported` property to a lower value.
- Fixed issue where `Generic6DOFJoint3D` and `JoltGeneric6DOFJoint3D` would yield odd limit shapes when using both linear and angular asymmetrical limits.
- Fixed issue where the equilibrium point for `Generic6DOFJoint3D` and `JoltGeneric6DOFJoint3D` would be moved when using asymmetrical limits.
- Fixed crash that could occur under rare circumstances when shutting down the editor after having added/removed collision shapes.
- Fixed issue where a `RigidBody3D` with locked axes colliding with a `StaticBody3D` (or another frozen `RigidBody3D` using `FREEZE_MODE_STATIC`) would result in NaNs.
- Fixed issue where `HingeJoint3D` and `JoltHingeJoint3D` would sometimes dull forces applied to either of its bodies when at either of its limits.
- Fixed issue with iOS `Info.plist` missing the `MinimumOSVersion` key.

<hr /><em>This discussion was created from the release <a href='https://github.com/godot-jolt/godot-jolt/releases/tag/v0.12.0-stable'>0.12.0-stable</a>.</em>

**Comment by ettiSurreal** [comment]
Are Cone shaped limits something that can be toggled/replicated? Cone shaped limits seem like they'd be very useful.

**Reply by mihe** [reply]
Not without building the extension yourself, no.

Cone-shaped angular limits come with the restriction of needing the upper/lower limits to always be symmetrical, which doesn't work all that well with how joints are configured in Godot. I had until now worked around this by rotating the reference frame for `node_a` internally to make the limits always be symmetrical, but that led to a bunch of issues, especially when using asymmetrical limits on multiple axes.

You might be able to just set these back to `JPH::ESwingType::Cone` to get the old behavior, but it's possible that you'll run into issues due to how property changes are applied in Godot:

https://github.com/godot-jolt/godot-jolt/blob/8ed0ce5b86dbb218645e5db927a19d14241a1ad4/src/joints/jolt_generic_6dof_joint_impl_3d.cpp#L515

https://github.com/godot-jolt/godot-jolt/blob/8ed0ce5b86dbb218645e5db927a19d14241a1ad4/src/joints/jolt_cone_twist_joint_impl_3d.cpp#L312
